include/zephyr/storage/storage_area/storage_area.h

```diff
@@ -0,0 +1,216 @@
+ * be a multiple of write_size and size should be a multiple of erase_size.
+ * The macro definitions xxxxx_storage_area(...) checks these conditions but
+ * always succeeds. Trying to use a badly sized storage area will result in
+ * failure of any of the exposed methods.
+ *
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+#define ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Storage_area interface
+ * @defgroup storage_area_interface Storage_area interface
+ * @ingroup storage_apis
+ * @{
+ */
+
+struct storage_area;
+
+struct storage_area_iovec { /* storage area io vector */
+	void *data;         /* pointer to data */
+	size_t len;         /* data length */
+};
+
+enum storage_area_properties_mask {
+	SA_PROP_READONLY = 0x0001,
+	SA_PROP_FOVRWRITE = 0x0002, /* full overwrite (ram, rram, ...) */
+	SA_PROP_LOVRWRITE = 0x0004, /* limited overwrite (nor flash) */
+	SA_PROP_ZEROERASE = 0x0008, /* erased value is 0x00 */
+	SA_PROP_AUTOERASE = 0x0010, /* erase while writing */
+};
+
+enum storage_area_ioctl_cmd {
+	SA_IOCTL_NONE,
+	SA_IOCTL_XIPADDRESS, /* retrieve the storage area xip address */
+};
+
+/**
+ * @brief storage_area API
+ *
+ * API to access storage area.
+ */
+struct storage_area_api {
+	int (*readv)(const struct storage_area *area, size_t start,
+		     const struct storage_area_iovec *iovec, size_t iovcnt);
+	int (*writev)(const struct storage_area *area, size_t start,
+		    const struct storage_area_iovec *iovec, size_t iovcnt);
+	int (*erase)(const struct storage_area *area, size_t start, size_t bcnt);
+	int (*ioctl)(const struct storage_area *area,
+		     enum storage_area_ioctl_cmd cmd, void *data);
+};
+
+/**
+ * @brief storage_area struct
+ */
+struct storage_area {
+	const struct storage_area_api *api;
+	size_t write_size;
+	size_t erase_size;
+	size_t erase_blocks;
+	uint32_t props; /* bitfield of storage area properties */
+};
+
+/**
+ * @brief storage_area macros
+ */
+#define STORAGE_AREA_HAS_PROPERTY(area, prop) ((area->props & prop) == prop)
+#define STORAGE_AREA_WRITESIZE(area)          area->write_size
+#define STORAGE_AREA_ERASESIZE(area)          area->erase_size
+#define STORAGE_AREA_SIZE(area)               area->erase_size * area->erase_blocks
+#define STORAGE_AREA_FOVRWRITE(area)                                            \
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_FOVRWRITE)
+#define STORAGE_AREA_LOVRWRITE(area)                                            \
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_LOVRWRITE)
+#define STORAGE_AREA_ERASEVALUE(area)                                           \
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_ZEROERASE) ? 0x00 : 0xff
+#define STORAGE_AREA_AUTOERASE(area)						\
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_AUTOERASE)
+
+/**
+ * @brief	 Read iovec from storage area.
+ *
+ * @param area   storage area.
+ * @param start  start in storage area (byte).
```

Comment:
This is a tricky subject - I would 💯 recommend not using `off_t` because it introduces a dependency cycle with POSIX.

But the other thing to consider is that `size_t` is usually limited to 32-bits on 32-bit systems, so the maximum size of an extent would be limited to 4GiB if `size_t` were used here.

Instead, it would be great if a `k_off_t` were defined with a size selectable via Kconfig.

Reasonable defaults would be for `k_off_t` to be 32-bits on 32-bit systems, and 64-bits on 64-bit systems. However, on 32-bit systems, users should be able to select 64-bits to be compatible with e.g. networked storage, large volumes, etc.

A new storage API would be a great justification to add `k_off_t`, but it would also get the ball rolling on fixing a bunch of tech-debt.